### PR TITLE
Add some helper structure to ease working with an item tree

### PR DESCRIPTION
**Pull request by hunger**

### CODE REVIEW

File: internal/core/item_tree.rs [hunk 1]
```diff
@@ -286,6 +286,108 @@ pub fn visit_item_tree<Base>(
     }
 }
 
+// Some helper to navigate the item_tree:
+
```
Comment (ogoffart) [inline internal/core/item_tree.rs:290]:
```suggestion
```

That comment is redundant with the next one. but not attached to anything.

File: internal/core/item_tree.rs [hunk 2]
```diff
@@ -286,6 +286,108 @@ pub fn visit_item_tree<Base>(
...
+/// This is a local item tree of one component with some helpful functions to navigate the tree.
+pub struct ItemTree<'a, T> {
+    item_tree: &'a [ItemTreeNode<T>],
+}
+
+impl<'a, T> ItemTree<'a, T> {
+    /// Cretae a new `ItemTree` from its raw data.
```
Comment (ogoffart) [inline internal/core/item_tree.rs:297]:
```suggestion
    /// Create a new `ItemTree` from its raw data.
```

File: internal/core/item_tree.rs [hunk 3]
```diff
@@ -286,6 +286,108 @@ pub fn visit_item_tree<Base>(
...
+    /// Get the parent of a node, returns `None` if this is the root node of this item tree.
+    pub fn parent(&self, index: usize) -> Option<usize> {
+        if index >= self.item_tree.len() || index == 0 {
+            None
+        } else {
+            match &self.item_tree[index] {
+                ItemTreeNode::Item { parent_index, .. } => Some(*parent_index as _),
+                ItemTreeNode::DynamicTree { parent_index, .. } => Some(*parent_index as _),
+            }
+        }
```
Comment (ogoffart) [inline internal/core/item_tree.rs:311]:
```suggestion
        (index < self.item_tree.len() && index != 0).then(|| self.item_tree[index].parent_index())
```
There is already a parent_index function.

File: internal/core/item_tree.rs [hunk 4]
```diff
@@ -286,6 +286,108 @@ pub fn visit_item_tree<Base>(
...
+            None // No parent, so we have no siblings either:-)
+        }
+    }
+
+    /// Returns the previous sibling or `None` if this is the first sibling.
+    pub fn previous_sibling(&self, index: usize) -> Option<usize> {
+        if let Some(parent_index) = self.parent(index) {
+            match self.item_tree[parent_index] {
+                ItemTreeNode::Item { children_count, .. } => {
+                    if index > children_count as usize {
```
Comment (ogoffart) [inline internal/core/item_tree.rs:339]: This looks wrong to me.
This shouldn't depends on children_count at all.
Maybe you mean the chioldren_index?

File: internal/core/item_tree.rs [hunk 5]
```diff
@@ -286,6 +286,108 @@ pub fn visit_item_tree<Base>(
...
+        }
+    }
+
+    /// Returns the first child or `None` if this are no children.
+    pub fn first_child(&self, index: usize) -> Option<usize> {
+        if index >= self.item_tree.len() {
+            return None;
+        }
+
+        match self.item_tree[index] {
```
Comment (ogoffart) [inline internal/core/item_tree.rs:360]:
```suggestion
        match self.item_tree.get(index)? {
```

File: internal/core/item_tree.rs [hunk 6]
```diff
@@ -286,6 +286,108 @@ pub fn visit_item_tree<Base>(
...
+            ItemTreeNode::Item { children_index, children_count, .. } => {
+                if children_count == 0 {
+                    None
+                } else {
+                    Some(children_index as _)
+                }
+            }
+            ItemTreeNode::DynamicTree { .. } => None,
```
Comment (ogoffart) [inline internal/core/item_tree.rs:368]: I don't know...  I guess this should be documented that the function return also None in this case.

File: internal/core/item_tree.rs [hunk 7]
```diff
@@ -286,6 +286,108 @@ pub fn visit_item_tree<Base>(
...
+        }
+    }
+
+    /// Returns the last child or `None` if this are no children.
+    pub fn last_child(&self, index: usize) -> Option<usize> {
+        if index >= self.item_tree.len() {
+            return None;
+        }
+
+        match self.item_tree[index] {
```
Comment (ogoffart) [inline internal/core/item_tree.rs:378]:
```suggestion
        match self.item_tree.get(index)? {
```

**Review by ogoffart — COMMENTED** [review]
In general, I'd rather see this kind of commit part of the PR that uses them.

I'm not sure it makes sense to have these function if they are not used.

File: internal/core/item_tree.rs (hunk 7 above)
Comment (hunger) [reply internal/core/item_tree.rs:378]: I will need to change this to not take ItemWeaks!

**Comment by hunger** [comment]
@ogoffart In general: I am trying to find some small hunks I can get reviewed individually so that the entire "tab focus" patch gets smaller.

File: internal/core/item_tree.rs (hunk 4 above)
Comment (hunger) [reply internal/core/item_tree.rs:330]: I think you are right! Thanks for catching this!

**Comment by hunger** [comment]
I'll push the entire thing in one PR...

**Pull request closed without merging**